### DISCUSSION TITLE
Additional data being sent and GSM module configuration

### DISCUSSION THREAD

**Discussion by aardrasystems** [Azure-Samples/iot-middleware-freertos-samples#240] [General]
hi @danewalton 

I hope all is well. 

I have a few additional questions.

Is the example project "sample_azure_iot_pnp_simulated_data" sending additional information to the IoT hub in addition to my message?

As I would like to only send a JSON string to the IoT hub, does this project send any information regards to twin data or config data?

The other question is it possible to configure the project to use a GSM module with ESP32? If I am correct we do not need to create a firmware that is based on AT commands, but just set the modem module in specific mode and project will use that method for communication instead of the wifi. Would this understanding be correct?

**Comment by danewalton-msft** [comment]
Okay looks like there's a few questions here.

> Sending a JSON

You can do this as telemetry or update properties in the device twin. If you send a device twin update, only the payload which you create will be sent. We do not send any other diagnostic information.

> GSM module

We do not currently have any support for that. There is a project being worked on for cellular support, but we do not have any updates on this at this time. If you wanted a more fine grained approach, [you can also use the Arduino repo for ESP32 and have more control over the MQTT and networking stack](https://github.com/Azure/azure-sdk-for-c-arduino).

**Comment by aardrasystems** [comment]
Hi @danewalton 

Thank you for your reply.

In regards to the GSM module, if I am correct, the free RTOS project depends on the ESP32 TCP/IP libraries when communicating and in theory, we would need to replace this with a GSM-based library?  Where in this project would I find the TCPIP integration to the free RTOS project? 


In regards to JSON, I am sending an updated JSON message once every minute or when a refresh data status is requested.  is this the correct approach to have this all in an infinite loop? Is there interrupt based option? My custom data is outputted from function sensor_data(),

```
 for( ; ; )
        {
            /* Hook for sending Telemetry */
         if ((message_tx_cycle_delay >= message_tx_cycle_delay_count) || (message_force_tx()==true))
            {

                        
                rst_message_force_tx();
                
                message_tx_cycle_delay = 0;
                if( ( ulCreateTelemetry( ucScratchBuffer, sizeof( ucScratchBuffer ), &ulScratchBufferLength ) == 0 ) &&
                    ( ulScratchBufferLength > 0 ) )
                {
                    
                    xResult = AzureIoTHubClient_SendTelemetry( &xAzureIoTHubClient,
                                                            ucScratchBuffer, ulScratchBufferLength,
                                                            NULL, eAzureIoTHubMessageQoS1, NULL );
                    configASSERT( xResult == eAzureIoTSuccess );
                }

                /* Hook for sending update to reported properties */
                ulReportedPropertiesUpdateLength = ulCreateReportedPropertiesUpdate( ucReportedPropertiesUpdate, sizeof( ucReportedPropertiesUpdate ) );

                if( ulReportedPropertiesUpdateLength > 0 )
                {
                    xResult = AzureIoTHubClient_SendPropertiesReported( &xAzureIoTHubClient, ucReportedPropertiesUpdate, ulReportedPropertiesUpdateLength, NULL );
                    configASSERT( xResult == eAzureIoTSuccess );
                }
            }
            else
            {
                message_tx_cycle_delay++;
            }
          //  LogInfo( ( "Attempt to receive publish message from IoT Hub.\r\n" ) );
            xResult = AzureIoTHubClient_ProcessLoop( &xAzureIoTHubClient,
                                                     sampleazureiotPROCESS_LOOP_TIMEOUT_MS );
            configASSERT( xResult == eAzureIoTSuccess );
            vTaskDelay( sampleazureiotDELAY_BETWEEN_PUBLISHES_TICKS );
        }


```

 ```
*/
uint32_t ulCreateTelemetry( uint8_t * pucTelemetryData,
                            uint32_t ulTelemetryDataSize,
                            uint32_t * ulTelemetryDataLength )
{


    int result = snprintf( ( char * ) pucTelemetryData, ulTelemetryDataSize,
                           sensor_data(), xDeviceCurrentTemperature );



    if( ( result >= 0 ) && ( result < ulTelemetryDataSize ) )
    {
        *ulTelemetryDataLength = result;
        result = 0;
    }
    else
    {
        result = 1;
    }

    return result;
}
```

**Reply by danewalton-msft** [reply]
Our sockets wrapper can be found here: 

- [FreeRTOS + TCPIP](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/demos/common/transport/sockets_wrapper_freertos_tcpip.c)
- [LWIP](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/demos/common/transport/sockets_wrapper_lwip.c)

As for the JSON, how to set that up is really up to you. As long as your device is operating within the limits of the Azure IoT Hub service and managing its MQTT connection okay, you should be alright. The code otherwise looks okay at quick glance.
